zos-utils/zossrc/AttlsContext.c

```diff
@@ -374,137 +393,178 @@ jboolean isQueryLoaded(JNIEnv *env, jobject obj)
     }
-    ioc->TTLSi_Req_Type = TTLS_QUERY_ONLY;
-    if (certificate) ioc->TTLSi_Req_Type |= TTLS_RETURN_CERTIFICATE;
 
-    ioc->TTLSi_BufferPtr = certificate ? (char*) getCertificateBuffer(env, obj) : (char*) NULL;
-    ioc->TTLSi_BufferLen = certificate ? buffer_certificate_size : 0;
+    c -> ioctl_buffer -> TTLSi_BufferPtr = c -> certificate_buffer;
+    c -> ioctl_buffer -> TTLSi_BufferLen = c -> certificate_buffer_length;
 
     // call ioctl
-    int rcIoctl = ioctl(getSocket(env, obj), SIOCTTLSCTL, (char*) ioc);
+    int rcIoctl = ioctl(c -> socket_id, SIOCTTLSCTL, c -> ioctl_buffer);
 
-    // if ioctl returns an error throw exception
     if (rcIoctl < 0) {
+        // if ioctl returns an error throw exception
         jclass exception_clazz = (*env) -> FindClass(env, JNI_CLASS_IOCTL_CALL_EXCEPTION);
         jmethodID constructor = (*env) -> GetMethodID(env, exception_clazz, JNI_METHOD_CONSTRUCTOR, JNI_SIGNATURE_METHOD_INT_INT_INT_VOID);
         jobject exception = (*env) -> NewObject(env, exception_clazz, constructor,
             (jint) rcIoctl, (jint) errno, (jint) __errno2());
         (*env) -> Throw(env, exception);
+    } else {
+        // update fields queryLoaded and certificateLoaded to avoid unnecessary call of IOCTL next time
+        (*env) -> SetBooleanField(env, obj, query_loaded_field, JNI_TRUE);
+        if (c -> load_certificate) {
+            (*env) -> SetBooleanField(env, obj, certificate_loaded_field, JNI_TRUE);
+        }
     }
 
-    return ioc;
+    return c;
 }
 
 /**
  * Return ioctl with data from query call. If data are available in memory, returns them, otherwise call ioctl.
  * If alwaysLoadCertificate is set to true certificated is also fetched.
  */
-struct TTLS_IOCTL* requireQuery(JNIEnv *env, jobject obj)
+Context* requireQuery(JNIEnv *env, jobject obj)
 {
-    if (isQueryLoaded(env, obj)) return getIoctl(env,obj);
-    return load(env, obj, JNI_FALSE);
+    if (isQueryLoaded(env, obj)) {
+        // query was done, just read the buffers
+        return getContext(env,obj, JNI_FALSE, JNI_FALSE);
+    }
+
+    // issue a new query with erased buffers
+    return query(env, obj, JNI_FALSE);
 }
 
 /**
  * Return ioctl with data with certificate. If data are available in memory, returns them, otherwise call ioctl.
  */
-struct TTLS_IOCTL* requireCertificate(JNIEnv *env, jobject obj)
+Context* requireCertificate(JNIEnv *env, jobject obj)
 {
-    if (isCertificateLoaded(env, obj)) return getIoctl(env,obj);
-    return load(env, obj, JNI_TRUE);
-}
-
-void cleanByteArray(JNIEnv *env, jbyteArray arr)
-{
-    if (!arr) return;
+    if (isCertificateLoaded(env, obj)) {
+        // query with certificate was done, just read the buffers
+        return getContext(env, obj, JNI_TRUE, JNI_FALSE);
+    }
 
-    (*env) -> CallStaticVoidMethod(env, arraysClass, arrays_fill_method_ID, arr, (jbyte) 0);
+    // issue a new query with erased buffers
+    return query(env, obj, JNI_TRUE);
 }
 
 /**
```

Comment:
I think a check for the `jobject obj` is not `NULL` is necessary. It is later used to call methods that have the `obj: a Java object (must not be NULL)` expectation.

Reply:
obj cannot be null, because it is only from instance methods. So yes, there is the non-null assumption